Update python guide (#90)
timescale/docs: timescaledb/quick-start/python.md

and running with TimescaleDB as their database. In this tutorial,
you'll learn how to:

> * [Connect to TimescaleDB](#connect-python-to-timescaledb)

Reviewer: I think we can skip the python part as we already have it in the URL. I love having the anchors too!
```suggestion
* [Connect to TimescaleDB](#connect-to-timescaledb)
```

> * [Connect to TimescaleDB](#connect-python-to-timescaledb)

Reviewer 2: Yeah, I agree the anchors are required. It would be good if we could come up with a way to make them easier to maintain, though.

> * [Connect to TimescaleDB](#connect-python-to-timescaledb)

Author: +1 for internal anchors. Do we have a way to make them work, as the old way which I deleted here doesn't work?

* [Create a relational table](#create-a-relational-table)
* [Create a hypertable](#create-hypertable)
* [Insert a batch of rows into TimescaleDB](#insert-rows-into-timescaledb)
* [Execute a query on TimescaleDB](#execute-a-query)

## Prerequisites

> 

Reviewer 2: Can we add a stem sentence in here? Something like:

```suggestion
Before you start, make sure you have:
```

> * At least cursory knowledge of Structured Query Language (SQL). The tutorial will walk you through each SQL command, 

Reviewer 2: ```suggestion
* At least some knowledge of SQL (structured query language). The tutorial will walk you through each SQL command, 
```

"some" just because it's a more commonly understood word. Switching the acronym/definition around because people are more likely to know "SQL" than the full spelled-out version.

The above method of composing a connection string is for test or development purposes only, for production purposes be sure to make sensitive details like your password, hostname, and port number environment variables.
</highlight>

### Step 3: Connect to TimescaleDB using the psycopg2 connect function

Use the psycopg2 [connect function][psycopg2-connect] to create a new database session and create 
a new [cursor object][psycopg2-cursor] to interact with the database.

In your `main` function, add the following lines:

```python
CONNECTION = "postgres://username:password@host:port/dbname"
def main():
    with psycopg2.connect(CONNECTION) as conn:
		cursor = conn.cursor()
        # use the cursor to interact with your database
        # cursor.execute("SELECT * FROM table")
```

Alternatively, you can create a connection object and pass the object
around as needed, like opening a cursor to perform database operations:

```python
CONNECTION = "postgres://username:password@host:port/dbname"
def main():  
    conn = psycopg2.connect(CONNECTION)
    cursor = conn.cursor()
    # use the cursor to interact with your database

>     # cursor.execute("SELECT * FROM table")

Reviewer: Maybe as the person is starting we can handle something like:
```suggestion
    print(cursor.execute("SELECT 'hello world'"))
```

Author: https://github.com/timescale/docs/pull/90/commits/981de17fa02761988313a37cbfba8f34670057e6

```

### Step 2: Execute the SQL statement and commit changes
Next, we execute the `CREATE TABLE` statement by opening a cursor, executing the
query from Step 1 and committing the query we executed in order to make the changes persistent. 
Afterward, we close the cursor to clean up:

```python
cursor = conn.cursor()
# see definition in Step 1
cursor.execute(query_create_sensors_table)
conn.commit()
cursor.close()
```

Congratulations, you've successfully created a relational table in TimescaleDB using Python.

## Create hypertable [](create_hypertable)

In TimescaleDB, the primary point of interaction with your data is a [hypertable][hypertable].
It provides an abstraction of a single continuous table across all space and time
intervals. You can can query it via standard SQL.

Virtually all user interactions with TimescaleDB are with hypertables. Creating tables
and indexes, altering tables, inserting data, selecting data, etc... can (and should)
all be executed on the hypertable.

A hypertable is defined by a standard schema with column names and types, with at

> least one column specifying a time value. [Read more about hypertables][hypertable-docs].

Reviewer 3: ```suggestion
least one column specifying a time value. [Learn more about using hypertables in the API documentation][hypertable-api].
```

```

> ### Step 2: Formulate create hypertable SELECT statement for your hypertable

Reviewer 3: ```suggestion
### Step 2: Formulate the SELECT statement to create your hypertable
```

While using `psycopg2` by itself may be sufficient for you to insert rows into your
hypertable, if you need quicker performance, you can use
[pgcopy][pg-copy-docs]. To do this, install `pgcopy` [using pip][pgcopy-install] and
then add this line to your list of `import` statements:

```python
from pgcopy import CopyManager
```

### Step 1: Get data to insert into database

> First we generate random sensor data - you would replace this step with funneling
> in your real data from your data pipeline.

Reviewer 3: I know we don't do this elsewhere, but since we don't specifically talk about `generate_series` and how this SQL works, it might be nice to at least say something like:

```suggestion
First we generate random sensor data using the `generate_series` function provided by PostgreSQL.
In the example query below, we will insert a total of 480 rows of data (4 readings, every 5 minutes, for 24 hours).
In your application, this would be the query that saves your time-series data into the hypertable.
in your real data from your data pipeline.
```

Reviewer 2: switch "we" for "you"

Author: https://github.com/timescale/docs/commit/eeb197b31ef0439ee0bcda1035199a3d0ccc93e4

    # create random data
    simulate_query = """SELECT generate_series(now() - interval '24 hour', now(), interval '5 minute') AS time,
                               %s as sensor_id,
                               random()*100 AS temperature,
                               random() AS cpu
                            """
    cursor.execute(simulate_query, data)
    values = cursor.fetchall()
```

### Step 2: Define columns of table you're inserting data into
Then we define the column names of the table we want to insert data into. In this
case, we're using the `sensor_data` hypertable that we created in the
"Generate a Hypertable" section above. This hypertable consists of the columns
named `time`, `sensor_id`, `temperature` and `cpu`. We define these column names
in a list of strings called `cols`.

```python
cols = ['time', 'sensor_id', 'temperature', 'cpu']
```

### Step 3: Instantiate a CopyManager with your target table and column definition
Lastly we create an instance of the `pgcopy` CopyManager, `mgr`, and pass our
connection variable, hypertable name, and list of column names. Then we use
the `copy` function of the CopyManager to insert the data into the database

> performantly using `pgcopy`.

Reviewer 2: Nit, but can we use "quickly" or "efficiently" here instead?

Reviewer 3: LOL. Totally missed this one. I don't mind "performant" (even though it's still not a real word and others hate it), but I've never seen this combination!

Author: https://github.com/timescale/docs/commit/bddb2c9871386cfb6dc3c2ac0cab4e1a8cd24c26

```python
# insert using pgcopy
def fast_insert(conn):
    cursor = conn.cursor()

    # for sensors with ids 1-4
    for id in range(1, 4, 1):
        data = (id,)
        # create random data
        simulate_query = """SELECT generate_series(now() - interval '24 hour', now(), interval '5 minute') AS time,
                           %s as sensor_id,
                           random()*100 AS temperature,
                           random() AS cpu
                        """
        cursor.execute(simulate_query, data)
        values = cursor.fetchall()

        # column names of the table you're inserting into
        cols = ['time', 'sensor_id', 'temperature', 'cpu']

        # create copy manager with the target table and insert
        mgr = CopyManager(conn, 'sensor_data', cols)
        mgr.copy(values)

    # commit after all sensor data is inserted
    # could also commit after each sensor insert is done
    conn.commit()
```

You can also check if the insertion worked:

```python
cursor.execute("SELECT * FROM sensor_data LIMIT 5;")
print(cursor.fetchall())
```

> Congratulations, you've successfully performantly inserted data into TimescaleDB

Reviewer 3: ```suggestion
Congratulations, you've successfully inserted time-series data into TimescaleDB
```

using Python and the `pgcopy` library.

> ## Execute a query [](execute_query)

Reviewer 3: So, totally NOT your fault, but it's weird that this tutorial jumps to a different dataset. And, now I'm realizing that all of these tutorials are a little bit different which is annoying and we'll work to standardize at some level.

Let's make this section about querying the data from what was just created above... the `sensor_data` table. Use the Golang tutorial to update it.

Reviewer 2: I really really want to re-write all of these tutorials 😬

Reviewer 3: Yeah, totally agree. I think we're close. Once we have one decent use case (like the prototypical "To-do app"), I'd like to give all of these a pass... and then we can work on the next example app, etc. Each should have something that's unique to help demonstrate features/uses.

Author: https://github.com/timescale/docs/commit/00ddcad84ce240d25d2f72b65148b8cb1ea13461

[continuous-aggregates]: /how-to-guides/continuous-aggregates/
[other-samples]: /tutorials/sample-datasets/
[migrate]: /how-to-guides/migrate-data/
[self-hosted-install]: http://localhost:8000/timescaledb/latest/how-to-guides/install-timescaledb/self-hosted/
[cloud-install]: https://www.timescale.com/timescale-signup
[pgcopy-install]: https://pypi.org/project/pgcopy/
[dictcursor-docs]: https://www.psycopg.org/docs/extras.html#dictionary-like-cursor

> [hypertable-docs]: /overview/core-concepts/hypertables-and-chunks/

Reviewer 3: ```suggestion
[hypertable]: /overview/core-concepts/hypertables-and-chunks/
[hypertable-api]: /api/:currentVersion:/hypertable/
```

Author: https://github.com/timescale/docs/pull/90/commits/d55f04b6511f59295d48bb66488ede6a850a0af2
